packages/custom_lint/lib/src/output.dart

```diff
@@ -0,0 +1,269 @@
+      .sortedBy((e) => _relativeFilePath(e.location.file, workingDirectory))
+      .sortedBy<num>((e) => -AnalysisErrorSeverity.VALUES.indexOf(e.severity));
+
+  if (errors.isEmpty) {
+    log.writeln('No issues found!');
+    return;
+  }
+
+  switch (format) {
+    case 'json':
+      emitJsonFormat(log, errors);
+      break;
+    default:
+      emitDefaultFormat(log, errors, workingDirectory);
+      break;
+  }
+}
+
+/// Emits a list of lints in the Dart analyzer style JSON format.
+@internal
+void emitJsonFormat(
+  Logger log,
+  Iterable<AnalysisError> errors,
+) {
+  Map<String, dynamic> location(String filePath, Map<String, dynamic> range) =>
+      {
+        'file': filePath,
+        'range': range,
+      };
+
+  Map<String, dynamic> position(int? offset, int? line, int? column) => {
+        'offset': offset,
+        'line': line,
+        'column': column,
+      };
+
+  Map<String, dynamic> range(
+    Map<String, dynamic> start,
+    Map<String, dynamic> end,
+  ) =>
+      {
+        'start': start,
+        'end': end,
+      };
+
+  final diagnostics = <Map<String, dynamic>>[];
+  for (final error in errors) {
+    final contextMessages = <Map<String, dynamic>>[];
+    if (error.contextMessages != null) {
+      for (final contextMessage in error.contextMessages!) {
+        final startOffset = contextMessage.location.offset;
+        contextMessages.add({
+          'location': location(
+            contextMessage.location.file,
+            range(
+              position(
+                startOffset,
+                contextMessage.location.startLine,
+                contextMessage.location.startColumn,
+              ),
+              position(
+                startOffset + contextMessage.location.length,
+                contextMessage.location.endLine,
+                contextMessage.location.endColumn,
+              ),
+            ),
+          ),
+          'message': contextMessage.message,
+        });
+      }
+    }
+    final startOffset = error.location.offset;
+    diagnostics.add({
+      'code': error.code,
+      'severity': error.severity,
+      'type': error.type,
+      'location': location(
+        error.location.file,
+        range(
+          position(
+            startOffset,
+            error.location.startLine,
+            error.location.startColumn,
+          ),
+          position(
+            startOffset + error.location.length,
+            error.location.endLine,
+            error.location.endColumn,
+          ),
+        ),
+      ),
+      'problemMessage': error.message,
+      if (error.correction != null) 'correctionMessage': error.correction,
+      if (contextMessages.isNotEmpty) 'contextMessages': contextMessages,
+      if (error.url != null) 'documentation': error.url,
+    });
+  }
+  log.writeln(json.encode({'version': 1, 'diagnostics': diagnostics}));
+}
+
+/// Emits a list of lints in the Dart analyzer style default format.
+@internal
+void emitDefaultFormat(
```

Comment:
Please don't change the output of the default mode in this PR.
That makes reviewing this PR a bit difficult

If we want to improve the rendering, this should be a separate issue/PR

Reply:
Yea that is basically most of the output file, I will try to split the `custom_lint` parts out and add a license-info/documentation at the top.

Reply:
But splitting up the json/default format into 2 PRs is kinda hard because they are pretty interwoven until the last print step.

Reply:
Sounds good then. I'm fine with having one PR if you move analyzer-stuff in its own file :)
Otherwise it's a bit hard to know what you wrote vs what's forked